### PR TITLE
Remove implicit types and attribute relations

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?
Implenting #5614, and **invalidates** the backwards compatibility of the DB data directory!
We split attribute ownerships from being full Grakn relations, using new storage structures: a simple edge for attribute key/ownerships, and in the schema `has` and `key` are new edge types. A major implication is that queries and reasoning over generic relations will no longer include attribute ownerships.

## What are the changes implemented in this PR?
* new Fragments - InHas and OutHas for traversing attribute ownerships separately from relations
* Remove `RelationStructure` and its subclasses, and integrate relations into the `Thing` hierarchy. This means old implicit relations are never accessible anymore!
* Refactor some class locations
* Remove `ImplicitType` from the `Schema.java` class!!!
* Change janus data structure to use `attribute` edges for attribute ownership, `has` and `key` edges for schema attribteu ownership/keyship, modify fragments and users correspondingly
* Remove `toRelationAtom()` in reasoner atoms - we can no longer treat attribute ownerships as relations and thus rules are less broadly applicable
* fix, update, or remove many tests

### CODE REVIEW

**Comment by flyingsilverfin** [comment]
cleanup TODOs:
* search and cleanup all usages of "reified"
* search for new TODOs and commentented code
* clean up comments on modified functions

**Comment by flyingsilverfin** [comment]
KEY validation at commit time is currently really slow - it validates `Thing` by checking it satisfies all keys it is supposed to have (we should be doing delta), and then for that key checks all the attribtue owners to see if any match the required type we're checking against.

may be moved to another issue for future work

**Comment by flyingsilverfin** [comment]
TODO: write a test to check that we don't infer appending role players

**Comment by flyingsilverfin** [comment]
TODO: implement ownership statistics as we can no longer store on edge vertices directly

**Comment by flyingsilverfin** [comment]
TODO: review the `SchemaMutationIT` and clean it up

File: concept/impl/RelationImpl.java [hunk 1]
```diff
@@ -18,124 +18,49 @@
...
-
-    /**
-     * Reads some data from a RelationReified. If the Relation has not been reified then an empty
-     * Stream is returned.
-     */
-    private <X> Stream<X> readFromReified(Function<RelationReified, Stream<X>> producer) {
-        RelationReified relationReified = reified();
-        return relationReified != null? producer.apply(relationReified) : Stream.empty();
-    }
+//    @Override
```
Comment (kasper-piskorski) [inline concept/impl/RelationImpl.java:60]: cleanup?

File: concept/manager/ConceptManagerImpl.java [hunk 2]
```diff
@@ -669,39 +615,42 @@ public LabelId convertToId(Label label) {
...
-        if (!transactionCache.isConceptCached(conceptId)) {
-            RelationImpl relation = new RelationImpl(new RelationEdge(edge, this, conceptNotificationChannel));
-            transactionCache.cacheConcept(relation);
-            return relation;
-        } else {
-            return transactionCache.getCachedConcept(conceptId);
-        }
-    }
+        throw new RuntimeException("No longer building edge concepts");
+//
```
Comment (kasper-piskorski) [inline concept/manager/ConceptManagerImpl.java:620]: commented out block

File: concept/structure/ElementFactory.java [hunk 3]
```diff
@@ -151,41 +147,41 @@ public VertexElement buildVertexElement(Vertex vertex) {
...
-                .has(Schema.EdgeProperty.RELATION_TYPE_LABEL_ID.name(), relationType.labelId().getValue())
-                .has(Schema.EdgeProperty.ROLE_LABEL_ID.name(), role.labelId().getValue())
-                .as("edge")
-                .inV()
-                .hasId(endVertexId)
-                .select("edge");
-
-        return traversal.hasNext();
-    }
+//    boolean rolePlayerEdgeExists(String startVertexId, Integer relationTypeLabelId, Integer roleLabelId, String endVertexId) {
```
Comment (kasper-piskorski) [inline concept/structure/ElementFactory.java:161]: ditto

File: concept/structure/ElementFactory.java [hunk 4]
```diff
@@ -225,17 +221,17 @@ boolean rolePlayerEdgeExists(String startVertexId, Type relationType, Role role,
...
-                .V()
-                .hasId(typeVertexId)
-                .in(Schema.EdgeLabel.SHARD.getLabel())
-                .in(Schema.EdgeLabel.ISA.getLabel())
-                .outE(Schema.EdgeLabel.ATTRIBUTE.getLabel())
-                .has(Schema.EdgeProperty.RELATION_TYPE_LABEL_ID.name(), edgeInstanceLabelId.getValue())
-                .toStream()
-                .map(edge -> buildEdgeElement(edge));
-    }
+//    Stream<EdgeElement> edgeRelationsConnectedToInstancesOfType(String typeVertexId, Integer edgeInstanceLabelId) {
```
Comment (kasper-piskorski) [inline concept/structure/ElementFactory.java:224]: ditto

File: concept/structure/VertexElementImpl.java [hunk 5]
```diff
@@ -240,32 +235,31 @@ public Shard shard() {
...
-        return elementFactory.rolePlayerEdges(id().toString(), type, allowedRoleTypeIds);
+    public Stream<EdgeElement> roleCastingsEdges(Integer typeLabelId, Set<Integer> allowedRoleTypeIds) {
+        return elementFactory.rolePlayerEdges(id().toString(), typeLabelId, allowedRoleTypeIds);
 
     }
-
-    public boolean rolePlayerEdgeExists(String startVertexId, RelationType type, Role role, String endVertexId) {
-        return elementFactory.rolePlayerEdgeExists(startVertexId, type, role, endVertexId);
-    }
+//
```
Comment (kasper-piskorski) [inline concept/structure/VertexElementImpl.java:242]: and here

File: graql/reasoner/atom/task/materialise/AttributeMaterialiser.java [hunk 6]
```diff
@@ -98,36 +95,36 @@ private ConceptMap findAnswer(Atom atom, ConceptMap sub, ReasoningContext ctx) {
...
+//    }
 
     /**
      * @param sub       partial substitution
      * @param owner     attribute owner
      * @param attribute attribute concept
      * @return inserted implicit relation if didn't exist, null otherwise
      */
-    private Relation putImplicitRelation(AttributeAtom atom, ConceptMap sub, Concept owner, Attribute attribute, ReasoningContext ctx) {
+    private void putImplicitRelation(AttributeAtom atom, ConceptMap sub, Concept owner, Attribute attribute, ReasoningContext ctx) {
```
Comment (kasper-piskorski) [inline graql/reasoner/atom/task/materialise/AttributeMaterialiser.java:124]: this needs renaming

File: graql/reasoner/rule/InferenceRule.java [hunk 7]
```diff
@@ -313,17 +312,17 @@ public boolean appendsRolePlayers(){
...
-            return new InferenceRule(
-                    reasonerQueryFactory.atomic(getHead().getAtom().toRelationAtom()),
-                    getBody(),
-                    rule,
-                    reasonerQueryFactory
-            );
-        }
-        return this;
-    }
+//    private InferenceRule rewriteHeadToRelation(Atom parentAtom){
```
Comment (kasper-piskorski) [inline graql/reasoner/rule/InferenceRule.java:315]: remove

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 8]
```diff
@@ -68,6 +61,187 @@
     //The ignored tests reveal some bugs in the reasoning algorithm, as they don't return the expected results,
     //as specified in the respective comments below.
 
+
```
Comment (kasper-piskorski) [inline test/integration/graql/reasoner/ReasoningIT.java:64]: cleanup needed

File: test/integration/graql/reasoner/atomic/RuleApplicabilityIT.java [hunk 9]
```diff
@@ -289,8 +287,10 @@ public void relationWithUnspecifiedRoles_roleplayersBoundByAttributes(){
             assertEquals(
                     rules.stream().filter(r -> r.getRule().label().getValue().contains("attributed-relation")).collect(toSet()),
                     relation3.getApplicableRules().collect(toSet()));
+
+            // exclude rules that infer attributes
             assertEquals(
-                    testTx.ruleCache().getRules().count() - 1,
+                    testTx.ruleCache().getRules().count() - 1 - 3,
```
Comment (kasper-piskorski) [inline test/integration/graql/reasoner/atomic/RuleApplicabilityIT.java:293]: ok, -1 -3 looks a bit magic now ;p

File: test/integration/graql/reasoner/query/MaterialisationIT.java [hunk 10]
```diff
@@ -154,22 +156,23 @@ public void whenMaterialisingAttributesThatAlreadyExist_noNewAttributesAreInsert
...
-
-            Statement attribute = Graql.var("x").has("resource-string", "materialised").id(entity.id().getValue());
-            Conjunction<Statement> pattern = Graql.and(Collections.singleton(attribute));
-            ReasonerAtomicQuery attributeQuery = reasonerQueryFactory.atomic(pattern);
-            ReasonerAtomicQuery implicitRelationQuery = reasonerQueryFactory.atomic(attributeQuery.getAtom().toRelationAtom());
-
-            List<ConceptMap> materialised = materialiseWithoutDuplicates(implicitRelationQuery.getPattern(),reasonerQueryFactory, tx);
-            assertTrue(tx.execute(attributeQuery.getQuery()).containsAll(materialised));
-        }
+//        try(Transaction tx = materialisationTestSession.transaction(Transaction.Type.WRITE)) {
```
Comment (kasper-piskorski) [inline test/integration/graql/reasoner/query/MaterialisationIT.java:162]: needs removing or updating

File: test/integration/graql/reasoner/query/MaterialisationIT.java [hunk 11]
```diff
@@ -202,18 +205,25 @@ public void whenMaterialisingAttributes_MaterialisedInformationIsCorrectlyFlagge
...
-                            "$r id " + resource.id().getValue() + ";" +
-                            "get;").asGet(), false)).get("rel").asRelation().isInferred());
+
+            Attribute<Object> attr = tx.getConcept(resource.id()).asAttribute();
+            List<Attribute<?>> owned = tx.getConcept(secondEntity.id()).asThing().attributes(attr.type())
+                    .filter(attributeOwned -> attributeOwned.equals(attr))
+                    .collect(Collectors.toList());
+            assertEquals(1, owned.size());
+
+//            assertTrue(Iterables.getOnlyElement(
```
Comment (kasper-piskorski) [inline test/integration/graql/reasoner/query/MaterialisationIT.java:215]: ditto

File: graql/reasoner/state/AtomicState.java [hunk 12]
```diff
@@ -167,13 +167,6 @@ private ConceptMap materialisedAnswer(ConceptMap baseAnswer, InferenceRule rule,
             RuleExplanation ruleExplanation = new RuleExplanation(Collections.singletonList(baseAnswer), rule.getRule());
             ConceptMap ruleAnswer = materialisedSub.explain(ruleExplanation);
             queryCache.record(ruleHead, ruleAnswer);
```
Comment (kasper-piskorski) [inline graql/reasoner/state/AtomicState.java:169]: 👌

File: server/session/TransactionImpl.java [hunk 13]
```diff
@@ -428,23 +430,56 @@ public Keyspace keyspace() {
...
                 .collect(Collectors.toSet());
 
-        ConceptUtils.getDependentConcepts(things)
-                .filter(Thing::isInferred)
-                .forEach(t -> {
-                    //as we are going to persist the concepts, reset the inferred flag
-                    ConceptVertex.from(t).vertex().property(Schema.VertexProperty.IS_INFERRED, false);
-                   transactionCache.inferredInstanceToPersist(t);
-                });
+        Set<Thing> visitedThings = new HashSet<>();
```
Comment (kasper-piskorski) [inline server/session/TransactionImpl.java:447]: 👌

**Pull request merged**